### DISCUSSION TITLE
What are the trade offs in using the cache directive

### DISCUSSION THREAD

**Discussion by akc42** [lit/lit#3580] [Q&A] [answered]
First off apologies for the enormous explanation below.  Some of It is necessary context for the question 
asked right at the end, but I also wanted to do a bit of "show and tell".  Since this is essentially a question
though I decided to put it here

I've been developing a Single Page Web App single handedly since 2013 originally written in JQuery. It is accompanied 
by an api server, originally written in PHP. Its for a customer I work part time for (I am otherwise 
retired) for replacing a Microsoft Access application that I have been supporting since 2010. Without 
being too specific and what this application does, the Access version is called PASv4 and the Web App is 
called PASv5.

In about 2016 with the appearance of Polymer I transitioned the client to pure Javascript and custom 
elements for the client and node.js for the api server.  As Lit and Lit2 came out I transitioned the 
client to that.  Operational custom elements (with a few exceptions) are called `<pas-xxxx>`

As of today, I have completed about 80 - 90% of the required functionality and as a result for the 
client have developed around 180 custom elements, with 8 mixins and 26 utility modules. I am starting
to worry that the browser may be struggling (I have no proof, but dev-tools frequently stops being 
able to show anything in the "elements" tab, and I also get complete crashes of the browser).

The structure of the custom elements is hierarchical, so for instance with `<pas-app>` being the only 
imported custom element in `index.html`, its shadow root contains a rendering of sub elements. The 
structure is something like the following (greatly simplified):-

```
                                              pas-app
                        				                 |
            -----------------------------------------------------------------------------
            |              |                     |                 |                 | | |
       pas-session     pas-error             pas-pages     pas-activity-monitor (various dialogs)
            |                                    |
        pas-logon         --------------------------------------------------------------------
                          |        |                 |              |         |        |  |  |
                      pas-menu pas-appointments pas-enquiries pas-customers pas-diary ...etc (dialogs)
                                   |                 |              |          |
                            -------------------     ...etc          |        ... etc
                            |                  |                    |
                 pas-appointment-month  pas-appointment-day         |
                                               |                    |
                                            ...etc                  |
                                                                    |
                                           -----------------------------------------------
                                           |                        |                    |
                                   pas-customer-details     pas-customer-finance   pas-questionnaires
```

In order to handle this spreads in the hierarchy I have a "page-manager" mixin which handles 
translation from urls (that is whole new discussion related to a distributed router I developed for the project ) to a "page" variable.  So the `render` function in `<pas-pages>` is of the
following form:-
```javascript
render() {
	return html`
	  ${cache({
	  	home: html`<pas-menu>Loading ...</pas-menu>`,
	  	appointments: html`<pas-appointments>Loading ...</pas-appointments>`,
	  	enquiries: html`<pas-enquiries>Loading ...</pas-enquiries>`,
	  	customers: html`<pas-customers>Loading ...</pas-customers>,
	  	diary: html`<pas-diary>Loading ...</pas-diary>`
	  }[this.page)}
	  <pas-confirmation-dialog></pas-confirmation-dialog>
	  <pas-time-dialog></dialog>
	  ${cache('appointments:diary'.indexOf(this.page) >= 0? html`
	  	<pas-staff-dialog></pas-staff-dialog>
	  `:'')}
	`
}
```

The use of the `{}[this.page]` construct allows different legs of the hierarchy to be included or excluded from the dom.

As an aside, you will notice that there are elements called `<pas-xxx-dialog>` These are generally 
singleton elements (extended from a `dialog-manager` mixin) which listen for a CustomEvent of the form
'pas-xxx-request' with any data they need in the event detail and open up (and can optionally position
themselves near the caller) when they receive the event.  When complete and not just cancelled, they
reply with `pas-xxx-reply` with any response data.  As a result they are placed as low in the hierarchy
consistent with all the calling elements being below them in the hierarchy also.

NOTE ALSO: The application is driven from data in a database and requested and updated by api calls.
There is not a single store on the client side (there would be far too much data), instead each key
element manages data appropriate to it.  So for instance `<pas-appointments>` fetches, holds and 
updates all appointments for the month.  However, its possible for instance for (say)
`<pas-customer-details>` to update the (again say) the customer name and that will need to be 
reflected in the appointments when that page is selected later.  Because `<pas-appointments>` is not
in the DOM (because of the paging mechanism above), I have had to introduce a memory based publish 
subscribe mechanism (called `keyupdate` and `register`), the code of which is as follows:-
```javascript
const callbackMap = new Map();  //This should be a map of entities that have been registered, with a callback array for each entity

export function registerUpdate(entity,callback) {
  let entities;
  if (typeof entity === 'string') {
    entities = [entity];
  } else if (Array.isArray(entity)) {
    entities = entity;
  } else {
    throw new Error('entity must be a string or an array of strings');
  }
  entities.forEach(entity => {
    if (typeof entity !== 'string') throw new Error('entities must be a string');
    let newEntry;
    if (callbackMap.has(entity)) {
      newEntry = callbackMap.get(entity);
      newEntry.push(callback);
    } else {
      newEntry = [callback];
    }
    callbackMap.set(entity, newEntry);
  });
}
export function keyUpdate(data) {
  let keystring = data.key;
  if (typeof keystring === 'object') {
    keystring = JSON.stringify(keystring);
  }
  debug(`update entity:${data.entity} key: ${keystring} action: ${data.action}`);
  const callbackArray = callbackMap.get(data.entity);
  if (typeof callbackArray !== 'undefined') callbackArray.forEach(callback => callback(data));
}
```
I wanted to put this in, because I saw a recent webinar from the Lit team about doing publish and subscribe using solely events and I wanted to show this because events don't always work.

NOTE: I do also have an element which is called like this 
`<pas-subsribe eventname="pas-break-lock-request" @pas-subscribed-event=${this._breakLock}>` 
which listens on "document.body" for the event in the eventname property/attribute and calls `this._breakLock` when an event 
reaches document.body. Its code is:-
```javascript
class PasSubscriber extends LitElement {
  static get styles() {
    return [css`
      :host {
        display: none;
      }
    `];
  }
  static get properties() {
    return {
      eventname: {type: String},
      private: {type: Boolean}  //if set we only listen for events sent to us, not subscribe in general
    };
  }
  constructor() {
    super();
    this.eventname = '';
    this.connected = false;
    this.private = false;
    this._reply = this._reply.bind(this);
  }
  connectedCallback() {
    super.connectedCallback();
    if (this.eventname.length > 0) {
      if (this.private) {
        this.addEventListener(this.eventname, this._reply);
      } else {
        document.body.addEventListener(this.eventname,this._reply);
      }
      this.connected = true;
    }

  }
  disconnectedCallback() {
    super.disconnectedCallback();
    if (this.connected) {
      this.connected = false;
      if (this.private) {
        this.removeEventListener(this.eventname, this._reply);
      } else {
        document.body.removeEventListener(this.eventname, this._reply);
      }
    }
  }
  update(changed) {
    if (changed.has('eventname')) {
      const oldname = changed.get('eventname')??'';
      if (oldname.length > 0 && this.connected) {
        this.connected = false;
        let wasPrivate = this.private;
        if (changed.has('private')) wasPrivate = !wasPrivate;
        if (wasPrivate) {
          this.removeEventListener(oldname, this._reply)
        } else {  
          document.body.removeEventListener(oldname, this._reply);
        }
      }
      if (this.eventname.length > 0 && !this.connected) {
        if (this.private) {
          this.addEventListener(this.eventname, this._reply);
        } else {
          document.body.addEventListener(this.eventname, this._reply);
        }
        this.connected = true;
      }
    }
    if(changed.has('private') && !changed.has('eventname')) {
      debug('alter listeners on private change');
      if(this.connected) {
        if (this.private) {
          document.body.removeEventListener(this.eventname, this._reply);
          this.addEventListener(this.eventname, this._reply);
        } else {
          this.removeEventListener(this.eventname, this._reply);
          document.body.addEventListener(this.eventname, this._reply);
        }
      }

    }
    super.update(changed);
  }
  _reply(e) {
    this.dispatchEvent(new CustomEvent('pas-subscribed-event',{detail:e.detail}));
  }
}
customElements.define('pas-subscriber',  PasSubscriber );
```  

I use this quite a lot for signaling events (as opposed to using variables) down the hierarchy. 

You will notice that the render function uses the `cache` directive twice, but I could just have easily left that out.  My early 
experience with Polymer led me to the tendency to use it, but lately I have tended not to.   But the big question is WHAT ARE THE TRADEOFFS in using it or not.

I can't quite figure out what the costs of using cache are.  What is cached of the hierarchy below the cache point, and what if it 
changes, does that mean a new version is cached as well?  For example the "customers" cause the html below that point to get cached, but what then what happens when the layer below changes from showing `pas-customer-details` to 
`pas-customer-finance`

By the same token, what re-render costs are saved by using cache in this example - does the whole hierarchy have to be completely  rendered from scratch or is it really only saving the immediate layer below.

**Comment by rictic** [comment, marked as answer]
The cache directive always uses more memory, and sometimes saves time.

When in doubt, benchmark. https://github.com/google/tachometer is a great tool for this.

When is it worth it? That depends on how much memory your users have to spare, and how expensive it is to re-render your specific templates from scratch. So that includes the constructors of your custom elements, how much work you do as part of a first update, and just how many dom nodes you make overall.

As a simple example of how cache works and compares against uncached, see this playground: https://lit.dev/playground/#gist=0ca161bc72c8c9a907540480840af61b

**Comment by akc42** [comment]
@rictic Thanks for that insight,  I'll try and benchmark some.  All my early developed modules had caching but I am starting to question the wisdom.  

The main answer I was trying to get though is this...  As you see from my diagram above, this a repeated hierarchy of branches of elements that are brought in an out of the dom.  If I cache(say) at the top level, what happens to a lower level uncached branch.  Are the various combinations of uncached branches held in the top level cache or are they still recreated on a render.

As you can imagine this offers two strategic options.  If the answer is the first of the two options I can see little point in caching further down the hierachy as it is already cached further up, if the answer is the second though, there is little point in caching at the top level of the hierarchy and I would be better off caching lower down (if I indeed need to).

**Reply by rictic** [reply]
It depends a lot on the specifics in your code, but caching at multiple layers in the tree will avoid doing HTML creation work.

The `cache` directive is keyed by template, so for example, in this case we'd expect the cache directive to improve perf:

```typescript
let template;
if (condition) {
  template = html`...expensive html...`;
} else {
  template = html`...alternate expensive html...`;
}
// creating the HTML will be cached
return html`${cache(template)}`;
```

This will not:


```typescript
let template;
if (condition) {
  template = html`...expensive html...`;
} else {
  template = html`...expensive alternative html...`;
}
// when 'condition' changes, the expensive html will be recreated not cached, because the cache is
// keyed by the template passed in, and this always passes in the same template (the one that just interpolates in
// the ${template} variable)
return html`${cache(html`${template}`)}`;
```

**Reply by akc42** [reply]
I get you .  Its a continuous tradeoff all the way down the tree.  each `${}` hole in the template which has some form of 
```
html` ... stuff...`
``` 
has to be recreated anyway unless it itself is an immediate condition of a cache directive
